File: server/internal/crypto/x3dh.go
package crypto

import (
	"crypto/ecdh"
	"fmt"
)

// x3dhInitiatorSecret implements Signal X3DH §3.3 without a one-time prekey:
//
//	DH1 = DH(IKa, SPKb)
//	DH2 = DH(EKa, IKb)
//	DH3 = DH(EKa, SPKb)
//	SK  = HKDF(F || DH1 || DH2 || DH3)
func x3dhInitiatorSecret(ika, eka *ecdh.PrivateKey, ikb, spkb []byte) ([]byte, error) {
	curve := ecdh.X25519()
	ikbPub, err := curve.NewPublicKey(ikb)
	if err != nil {
		return nil, fmt.Errorf("x3dh: ikb: %w", err)
	}
	spkbPub, err := curve.NewPublicKey(spkb)
	if err != nil {
		return nil, fmt.Errorf("x3dh: spkb: %w", err)
	}
	dh1, err := ika.ECDH(spkbPub)
	if err != nil {
		return nil, err
	}
	dh2, err := eka.ECDH(ikbPub)
	if err != nil {
		return nil, err
	}
	dh3, err := eka.ECDH(spkbPub)
	if err != nil {
		return nil, err
	}
	return x3dhKDF(dh1, dh2, dh3)
}

// x3dhResponderSecret is the Bob-side of the same three DHs.
func x3dhResponderSecret(ikb, spkb *ecdh.PrivateKey, ika, eka []byte) ([]byte, error) {
	curve := ecdh.X25519()
	ikaPub, err := curve.NewPublicKey(ika)
	if err != nil {
		return nil, fmt.Errorf("x3dh: ika: %w", err)
	}
	ekaPub, err := curve.NewPublicKey(eka)
	if err != nil {
		return nil, fmt.Errorf("x3dh: eka: %w", err)
	}
	dh1, err := spkb.ECDH(ikaPub)
	if err != nil {
		return nil, err
	}
	dh2, err := ikb.ECDH(ekaPub)
	if err != nil {
		return nil, err
	}
	dh3, err := spkb.ECDH(ekaPub)
	if err != nil {
		return nil, err
	}
	return x3dhKDF(dh1, dh2, dh3)
}

func x3dhKDF(dh1, dh2, dh3 []byte) ([]byte, error) {
	ikm := make([]byte, 0, 32+32*3)
	ikm = append(ikm, x3dhF...)
	ikm = append(ikm, dh1...)
	ikm = append(ikm, dh2...)
	ikm = append(ikm, dh3...)
	return hkdfSHA256(ikm, zeros32, []byte(InfoX3DH), 32)
}

func associatedData(initiatorIK, responderIK []byte) []byte {
	ad := make([]byte, 0, 64)
	ad = append(ad, initiatorIK...)
	ad = append(ad, responderIK...)
	return ad
}

File: ios/Faraday/Faraday/Crypto/X3DH.swift
import CryptoKit
import Foundation

enum X3DH {
    static func initiatorSecret(
        ika: Curve25519.KeyAgreement.PrivateKey,
        eka: Curve25519.KeyAgreement.PrivateKey,
        ikb: Data,
        spkb: Data
    ) throws -> Data {
        let ikbPub = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: ikb)
        let spkbPub = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: spkb)
        let dh1 = try ika.sharedSecretFromKeyAgreement(with: spkbPub).data
        let dh2 = try eka.sharedSecretFromKeyAgreement(with: ikbPub).data
        let dh3 = try eka.sharedSecretFromKeyAgreement(with: spkbPub).data
        return try kdf(dh1, dh2, dh3)
    }

    static func responderSecret(
        ikb: Curve25519.KeyAgreement.PrivateKey,
        spkb: Curve25519.KeyAgreement.PrivateKey,
        ika: Data,
        eka: Data
    ) throws -> Data {
        let ikaPub = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: ika)
        let ekaPub = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: eka)
        let dh1 = try spkb.sharedSecretFromKeyAgreement(with: ikaPub).data
        let dh2 = try ikb.sharedSecretFromKeyAgreement(with: ekaPub).data
        let dh3 = try spkb.sharedSecretFromKeyAgreement(with: ekaPub).data
        return try kdf(dh1, dh2, dh3)
    }

    private static func kdf(_ dh1: Data, _ dh2: Data, _ dh3: Data) throws -> Data {
        var ikm = FaradayKDF.x3dhF
        ikm.append(dh1)
        ikm.append(dh2)
        ikm.append(dh3)
        return try FaradayKDF.hkdf(ikm: ikm, salt: FaradayKDF.zeros32, info: Data(FaradayKDF.infoX3DH.utf8), length: 32)
    }

    static func associatedData(initiatorIK: Data, responderIK: Data) -> Data {
        var ad = initiatorIK
        ad.append(responderIK)
        return ad
    }
}

extension SharedSecret {
    var data: Data {
        withUnsafeBytes { Data($0) }
    }
}
